src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs

```diff
@@ -245,56 +261,64 @@ private static void RecordRequestStartEventLog(HttpContext httpContext)
         }
 
         [MethodImpl(MethodImplOptions.NoInlining)]
-        private Activity StartActivity(HttpContext httpContext, out bool hasDiagnosticListener)
+        private Activity? StartActivity(HttpContext httpContext, out bool hasDiagnosticListener)
         {
-            var activity = new Activity(ActivityName);
             hasDiagnosticListener = false;
+            if (_diagnosticListener.IsEnabled(ActivityStartKey))
+            {
+                hasDiagnosticListener = true;
+            }
+
+            // Short-circuit to avoid doing an expensive header lookup
+            if (!(hasDiagnosticListener || _activitySource.HasListeners()))
+            {
+                return null;
+            }
 
             var headers = httpContext.Request.Headers;
             if (!headers.TryGetValue(HeaderNames.TraceParent, out var requestId))
             {
                 headers.TryGetValue(HeaderNames.RequestId, out requestId);
             }
-
-            if (!StringValues.IsNullOrEmpty(requestId))
+            var activity = _activitySource.StartActivity(ActivityName, ActivityKind.Server, requestId);
```

Comment:
>requestId [](start = 92, length = 9)

Just to confirm, does the requestId is well formed W3C id? I am asking because if it is not then the created listener wouldn't allow creating the Activity because the SampleUsingParentId callback is not provided.

Reply:
There are no guarantee. It is unsanitized user input from an incoming HTTP header.

I'm assuming I should validate?

Reply:
The question is, do you need to create the Activity when getting non W3C ids? if so, then you may need to add SampleUsingParentId to the listener too.

Reply:
I'd like to preserve compat. We currently create a new Activity even when receiving non-W3C ids.

Reply:
Good. so please add `SampleUsingParentId` to the listener.

Reply:
Do I need both? Assuming my requestId in w3c format, will just SampleUsingParentId suffice?

Reply:
SampleUsingParentId should be sufficient. `Sample` would be needed only in case creating the Activity using `ActivityContext`.

Reply:
Turns out I do need both. SampleUsingParentId doesn't fire when there is no parentId present 🤦🏾‍♂️

Reply:
That is right when parentId = `null`. I thought you would always have non-null parent id but it looks this not the case. so, you need both.